src/include/storage/write_ahead_log/log_io.h

```diff
@@ -149,11 +162,44 @@ class BufferedLogWriter {
    */
   bool IsBufferFull() { return buffer_size_ == common::Constants::LOG_BUFFER_SIZE; }
 
+  /**
+   * Mark that the BufferedLogWriter is now ready for serialization.
+   *
+   * Currently, BufferedLogWriter is used to serialize both over the disk and over the network.
+   * To avoid copying, we refcount the serialization usages of the BufferedLogWriter.
+   * This function initializes the refcount.
+   *
+   * TODO(WAN): Come up with a better way of "sharing" buffers?
+   */
+  void PrepareForSerialization() {
+    NOISEPAGE_ASSERT(serialize_refcount_.load() == 0, "This buffer is already being serialized.");
+    serialize_refcount_.store(NUM_SERIALIZERS);
+  }
+
+  /**
+   * Mark one successful serialization of the buffered log.
+   *
+   * This should be called exactly once for each serializer of this log. See PrepareForSerialization() for more info.
+   * @return True if the current log has been completely serialized, meaning that no serializers are left and that
+   *         it is safe to now reuse this BufferedLogWriter.
+   */
+  bool MarkSerialized() {
+    auto count = serialize_refcount_.fetch_sub(1);
+    NOISEPAGE_ASSERT(serialize_refcount_.load() >= 0, "This buffer was serialized too many times?");
+    return count == 1;
+  }
+
  private:
+  friend class replication::ReplicationManager;
+
+  /** The number of callers of MarkSerialized(). See also PrepareForSerialization(). */
+  static constexpr int NUM_SERIALIZERS = 2;
```

Comment:
Might be helpful to include in the comment who the callers are

Reply:
Removed.